### PR TITLE
Bugfix for value removals from read only array container

### PR DESCRIPTION
**Pull request by xtonik**
### SUMMARY
There was bad loop iteration, which caused all consecutive values are omitted from results of actions `flip()` of present value and `remove()` on array container backed by read only buffer.

Creation of mutable copy of backing buffer is necessary as otherwise the exception `ReadOnlyBufferException` was thrown, just run `TestMappeableArrayContainer` after checkout 556cfe3.

Condition to increase capacity was omitted as it is always satisfied - I have not found any way to create not read only heap char buffer, am I wrong?

Note, that misconception between direct array and not read only buffer is widely spread across whole project, when recognition between all variants are done by `isBackedBySimpleArray()`, but nowadays I suppose it is better left as it is as this is strongly breakable change. I mean there are three variants at least:
1. mutable direct array buffer - do everything what you need
2. mutable indirect access buffer - you can change the values, but you have to iterate one by one through the values
3. read only buffer whatsoever backed by array or some indirect access buffer - you have to iterate one by one, if you want to do some changes, mutable buffer copy must be done

### Automated Checks

- [x] I have run `./gradlew test` and made sure that my PR does not break any unit test.
- [x] I have run `./gradlew checkstyleMain` or the equivalent and corrected the formatting warnings reported.

### CODE REVIEW

File: RoaringBitmap/src/main/java/org/roaringbitmap/buffer/MappeableArrayContainer.java [hunk 1]
```diff
@@ -521,9 +513,15 @@ public MappeableContainer flip(char x) {
         content.put(-loc - 1, x);
         ++cardinality;
       } else {
-        for (int k = loc + 1; k < cardinality; --k) {
-          content.put(k - 1, content.get(k));
+        // buffer is read only, we must make mutable copy
```
Comment (lemire) [inline RoaringBitmap/src/main/java/org/roaringbitmap/buffer/MappeableArrayContainer.java:516]: This is not relevant to this PR. Please add it to a separate PR.

Further, if you think that it is necessary, please demonstrate using our main API.

File: RoaringBitmap/src/main/java/org/roaringbitmap/buffer/MappeableArrayContainer.java [hunk 2]
```diff
@@ -1408,14 +1406,18 @@ public MappeableContainer remove(final char x) {
     } else {
       final int loc = BufferUtil.unsignedBinarySearch(content, 0, cardinality, x);
       if (loc >= 0) {
-        // insertion
-        for (int k = loc + 1; k < cardinality; --k) {
-          content.put(k - 1, content.get(k));
+        // buffer is read only, we must make mutable copy
```
Comment (lemire) [inline RoaringBitmap/src/main/java/org/roaringbitmap/buffer/MappeableArrayContainer.java:1409]: This is not relevant to this PR. Please add it to a separate PR.

File: RoaringBitmap/src/test/java/org/roaringbitmap/buffer/TestMappeableArrayContainer.java [hunk 3]
```diff
@@ -68,6 +68,33 @@ public void remove() {
     assertTrue(ac.contains((char) 1));
   }
 
+  @Test
+  public void removePresentValueFromReadOnlyContainer() {
```
Comment (lemire) [inline RoaringBitmap/src/test/java/org/roaringbitmap/buffer/TestMappeableArrayContainer.java:72]: Please write similar tests in TestArrayContainer.java.

https://github.com/RoaringBitmap/RoaringBitmap/blob/master/RoaringBitmap/src/test/java/org/roaringbitmap/TestArrayContainer.java

We try to have the same test coverage. If you found a bug in MappeableArrayContainer.java, there might be one in ArrayContainer.java.

**Comment by lemire** [comment]
> Note, that misconception between direct array and not read only buffer is widely spread across whole project, when recognition between all variants are done by isBackedBySimpleArray(), but nowadays I suppose it is better left as it is as this is strongly breakable change. I mean there are three variants at least: mutable direct array buffer - do everything what you need, mutable indirect access buffer - you can change the values, but you have to iterate one by one through the values, read only buffer whatsoever backed by array or some indirect access buffer - you have to iterate one by one, if you want to do some changes, mutable buffer copy must be done

This should be a separate PR.

Further, motivate this use case with actual code using our public API (MutableRoaringBitmap and ImmutableRoaringBitmap). Please don't code functionality within the containers that may not be related to the functionality that our users rely upon.

**Comment by lemire** [comment]
I cannot seem to verify the issue you are reporting:

https://github.com/RoaringBitmap/RoaringBitmap/pull/677

Please provide evidence, using our public API, that there is a bug.

*Do not change the internals of the library without motivation. I will not merge speculative PRs.*

**Comment by xtonik** [comment]
As stated above this PR is completely irrelevant as it is not part of API and thus responsibility lies on calling user.

**Pull request closed without merging**